demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt

```diff
@@ -1,43 +1,116 @@
 
 private const val TAG = "FederatedCycleViewModel"
+private const val EPOCHS = 1
 
 @ExperimentalUnsignedTypes
+@ExperimentalStdlibApi
 class FederatedCycleViewModel(
     baseUrl: String,
     authToken: String,
+    private val mnistDataRepository: MNISTDataRepository,
     networkSchedulers: ProcessSchedulers,
-    computeSchedulers: ProcessSchedulers
+    computeSchedulers: ProcessSchedulers,
+    private val localConfiguration: LocalConfiguration
 ) : ViewModel() {
     private val syftWorker = Syft.getInstance(
         baseUrl, authToken,
         networkSchedulers, computeSchedulers
     )
-    private val mnistJob = syftWorker.newJob("mnist", "1.0.0")
+    private val mnistJob = syftWorker.newJob("mnist", "1.0.3")
+
+    val logger
+        get() = _logger
+    private val _logger = MutableLiveData<String>()
+
+    val processState
+        get() = _processState
+    private val _processState = MutableLiveData<ProcessState>()
+
+    val processData
+        get() = _processData
+    private val _processData = MutableLiveData<ProcessData>()
 
     fun startCycle() {
-        Log.d(TAG, "mnist job started")
+        postLog("MNIST job started")
+        postState(ProcessState.Loading)
         val jobStatusSubscriber = object : JobStatusSubscriber() {
-            override fun onReady(model: SyftModel, clientConfig: ClientConfig) {
-                //todo training code goes here
+
+            override fun onReady(
+                model: SyftModel,
+                plans: ConcurrentHashMap<String, Plan>,
+                clientConfig: ClientConfig
+            ) {
+                postLog("Model ${model.modelName} received.\nStarting training process")
+                trainingProcess(model, plans, clientConfig)
             }
 
             override fun onRejected(timeout: String) {
-                //TODO retry after timeout
+                postLog("We've been rejected $timeout")
             }
 
             override fun onError(throwable: Throwable) {
-                //todo error functionality
+                postLog("There was an error $throwable")
             }
         }
         mnistJob.start(jobStatusSubscriber)
     }
-}
\ No newline at end of file
+
+    private fun trainingProcess(
+        model: SyftModel,
+        plans: ConcurrentHashMap<String, Plan>,
+        clientConfig: ClientConfig
+    ) {
+
+        plans.values.first().let { plan ->
+            val result = mutableListOf<Float>()
+            val loadData = mnistDataRepository.loadData()
+            val zipData = loadData.first zip loadData.second
+            repeat(EPOCHS) {
+                postLog("Starting epoch ${it + 1}")
+                zipData.forEach { trainingBatch ->
+                    val output = plan.execute(
+                        model,
+                        trainingBatch,
+                        clientConfig.copy(batchSize = 1)
+                    )?.toTuple()
+                    output?.let { outputResult ->
```

Comment:
I ran the demo and everything looks great but can you populate the graph dynamically after say every 10 epochs rather than waiting till the end for training to finish. I would like to see if the loss error actually reduces over time. Also is the training data the complete MNIST data?